storage/innobase/log/log0log.cc

```diff
@@ -1023,27 +1034,28 @@ template<bool release_latch> inline lsn_t log_t::write_buf() noexcept
         (We want to avoid memset() while holding exclusive log_sys.latch)
         This block will be overwritten later, once records beyond
         the current LSN are generated. */
-#ifdef HAVE_valgrind
-          MEM_MAKE_DEFINED(buf + length, (write_size_1 + 1) - new_buf_free);
-        if (UNIV_LIKELY_NULL(re_write_buf))
-          MEM_MAKE_DEFINED(re_write_buf + length, (write_size_1 + 1) -
-                           new_buf_free);
-#endif
+        MEM_MAKE_DEFINED(buf + length, (write_size_1 + 1) - new_buf_free);
         buf[length]= 0; /* allow recovery to catch EOF faster */
         length&= ~write_size_1;
         memcpy_aligned<16>(flush_buf, buf + length, (new_buf_free + 15) & ~15);
-        if (UNIV_LIKELY_NULL(re_write_buf))
+        if (resizing < 0 ? UNIV_UNLIKELY(!!re_write_buf) : bool(resizing))
+        {
+          MEM_MAKE_DEFINED(re_write_buf + length, (write_size_1 + 1) -
+                           new_buf_free);
           memcpy_aligned<16>(resize_flush_buf, re_write_buf + length,
                              (new_buf_free + 15) & ~15);
```

Comment:
> if (resizing < 0 ? UNIV_UNLIKELY(!!re_write_buf) : bool(resizing))

In all three places we have this condition, it should be safe and better to keep the old condition i.e. checking re_write_buf directly.
1. resizing < 0 : already has UNIV_UNLIKELY(!!re_write_buf)
2. resizing >=0 : byte *const re_write_buf{resizing ? resize_buf : nullptr};
Also, if re_write_buf is nullptr, certainly the code below is going to crash. So, I am not sure why not use
> if (UNIV_LIKELY_NULL(re_write_buf))

Reply:
Your suggestion to just check `re_write_buf` without any additional template condition would cause an unnecessary runtime condition for the case `resizing=1` where we know that resizing is in progress. This is because the local variable `re_write_buf` is being initialized as `resizing ? resize_buf : nullptr` and the compiler can’t know that the pointer is not supposed to be `nullptr`.

Earlier in `log_t::write_buf()` there are two debug assertions for ensuring that the compile-time and run-time parameters match:
```c++
  ut_ad(resizing < 0 || bool(resizing) == (resize_in_progress() > 1));
// …
    byte *const re_write_buf{resizing ? resize_buf : nullptr};
    ut_ad(resizing < 0 || bool(resizing) == !!re_write_buf);
```
I just realized that we could possibly use precursors of the [C++23 `assume`](https://en.cppreference.com/w/cpp/language/attributes/assume), like we do elsewhere:
```c++
#ifdef _MSC_VER
template <std::size_t Alignment, class T>
static inline T my_assume_aligned(T ptr)
{
  assert(reinterpret_cast<size_t>(ptr) % Alignment == 0);
  __assume(reinterpret_cast<size_t>(ptr) % Alignment == 0);
  return ptr;
}
#else
template <std::size_t Alignment, class T>
static inline T my_assume_aligned(T ptr)
{
  assert(reinterpret_cast<size_t>(ptr) % Alignment == 0);
  return static_cast<T>(__builtin_assume_aligned(ptr, Alignment));
}
#endif
```
I will try that.

Reply:
The `__builtin_unreachable()` based solution that I implemented in c391fb1ff122adbfb9a7f98d3239aa08311918d6 seems to work, both in `CMAKE_BUILD_TYPE=RelWithDebInfo` and in optimized `CMAKE_BUILD_TYPE=Debug`:
```asm
   call   0xd69270 <log_write_buf(byte const*, size_t, lsn_t)>
   mov    %rbx,%rdx
   mov    %r14,%rsi
   lea    0x1ae4750(%rip),%rdi        # 0x2186d00 <log_sys>
   call   0xd69ca0 <_ZN5log_t16resize_write_bufEPKhm>
```
There is no conditional branch between the two `call` near the end of the function `log_writer_resizing()`. The `if` condition in the following is essentially being replaced with `true`:
```c++
    log_write_buf(write_buf, length, offset);

    if (UNIV_LIKELY_NULL(re_write_buf))
      resize_write_buf(re_write_buf, length);
```